cpp/DEVELOPER_GUIDE.md

```diff
@@ -5,9 +5,22 @@ This document summarizes rules and best practices for contributions to the cuML
 Please start by reading [CONTRIBUTING.md](../CONTRIBUTING.md).
 
 ## Thread safety
-cuML is thread safe so its functions can be called from multiple host threads if they use different handles.
 
-The implementation of cuML is single threaded.
+With the exception of the cumlHandle, cuML algorithms should maintain thread-safety and are, in general, 
+assumed to be single threaded. This means they should be able to be called from multiple host threads so 
+long as different instances of `cumlHandle` are used.
+
+Exceptions are made for algorithms that can take advantage of multiple CUDA streams in order to oversubscribe 
```

Comment:
One can use multiple CUDA streams from a single CPU thread so better to refer to CPU threads here.